### PR TITLE
Added Docstrings to :class:`~.Mobject`

### PR DESCRIPTION
**Pull request by kolibril13**
As there are still a lot of undocumented functions, this pr will add some more docstings to mobject methods (e.g. rotate, get_x, flip, etc.
EDIT: SEE rendered page here: https://manimce--1206.org.readthedocs.build/en/1206/reference/manim.mobject.mobject.Mobject.html
## Checklist
- [x] I have read the [Contributing Guidelines](https://docs.manim.community/en/latest/contributing.html)
- [x] I have written a descriptive PR title (see top of PR template for examples)
- [x] I have written a changelog entry for the PR or deem it unnecessary
- [x] My new functions/classes either have a docstring or are private
- [x] My new functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [x] My new documentation builds, looks correctly formatted, and adds no additional build warnings

## Reviewer Checklist
- [ ] The PR title is descriptive enough
- [ ] The PR is labeled correctly
- [ ] The changelog entry is completed if necessary
- [ ] Newly added functions/classes either have a docstring or are private
- [ ] Newly added functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [ ] Newly added documentation builds, looks correctly formatted, and adds no additional build warnings

### CODE REVIEW

**Comment by kolibril13** [comment]
I am just wondering:
https://docs.manim.community/en/latest/examples.html#rotationupdater
has a reference to `add_updater()`but it is not possible to click on it, however there is this page that gives an example how to use updater:
https://docs.manim.community/en/latest/reference/manim.mobject.mobject.Mobject.html#manim.mobject.mobject.Mobject.add_updater

**Comment by jsonvillanueva** [comment]
I figured out the issue, it's because we need our manim_directive.py has :ref_functions: link to :func:`~.{func_here}`... but the issue with this is that these aren't functions. They are methods. As such we need a new directive called :meth:`~.{meth_here}`

**Comment by kolibril13** [comment]
https://github.com/ManimCommunity/manim/blob/c3f6119761a10991d45cbd46b218c9dc54d037e1/docs/source/manim_directive.py#L135
Found it!
ok, I can add this to this pr.

File: manim/mobject/mobject.py [hunk 1]
```diff
@@ -1721,33 +1724,39 @@ def length_over_dim(self, dim):
...
     def get_x(self, direction=ORIGIN):
+        """Returns x coordinate of the center of the ``mobject`` as ``numpy.float64`` """
         return self.get_coord(0, direction)
 
     def get_y(self, direction=ORIGIN):
+        """Returns y coordinate of the center of the ``mobject`` as ``numpy.float64`` """
         return self.get_coord(1, direction)
 
     def get_z(self, direction=ORIGIN):
+        """Returns z coordinate of the center of the ``mobject`` as ``numpy.float64`` """
```
Comment (kolibril13) [inline manim/mobject/mobject.py:1739]: should this be shorter?
At the moment, I think it looks a bit to long: 
image

Comment (jsonvillanueva) [reply manim/mobject/mobject.py:1739]: > 
> 
> https://github.com/ManimCommunity/manim/blob/c3f6119761a10991d45cbd46b218c9dc54d037e1/docs/source/manim_directive.py#L135
> 
> Found it!
> ok, I can add this to this pr.

I just made #1208, feel free to do it here if you wish, but I'd still suggest a separate PR if this one is supposed to be for documentation/docstrings

File: manim/mobject/mobject.py [hunk 2]
```diff
@@ -1011,9 +1011,11 @@ def scale(self, scale_factor: float, **kwargs) -> "Mobject":
             return self
 
     def rotate_about_origin(self, angle, axis=OUT, axes=[]):
+        """Rotates an mobject about the ORIGIN, which is at [0,0,0]"""
         return self.rotate(angle, axis, about_point=ORIGIN)
 
     def rotate(self, angle, axis=OUT, **kwargs):
+        """Rotates an mobject about a certain point e.g. `m.rotate(PI, about_point = 2*LEFT)`"""
```
Comment (kolibril13) [inline manim/mobject/mobject.py:1018]: The last part is here not rendered in the docs:
image
Do you know why?
Maybe to long?

Comment (kolibril13) [reply manim/mobject/mobject.py:1018]: I'll try that.

File: manim/mobject/mobject.py [hunk 3]
```diff
@@ -1011,9 +1011,11 @@ def scale(self, scale_factor: float, **kwargs) -> "Mobject":
             return self
 
     def rotate_about_origin(self, angle, axis=OUT, axes=[]):
+        """Rotates an mobject about the ORIGIN, which is at [0,0,0]"""
         return self.rotate(angle, axis, about_point=ORIGIN)
 
     def rotate(self, angle, axis=OUT, **kwargs):
+        """Rotates an mobject about a certain point e.g. ``m.rotate(PI, about_point = 2*LEFT)``"""
```
Comment (kolibril13) [inline manim/mobject/mobject.py:1018]: This is a bit inconvenient:
image
but later the code is rendered:
image
So should I remove the code ``m.rotate(PI, about_point = 2*LEFT)`` completely or make a new line?

Comment (behackl) [reply manim/mobject/mobject.py:1018]: The oneline summary is not intended for examples. If you want to add an example, do so in a new line (preferably as a doctest then).

File: manim/mobject/mobject.py [hunk 4]
```diff
@@ -1721,33 +1750,39 @@ def length_over_dim(self, dim):
...
     def get_y(self, direction=ORIGIN):
+        """Returns y coordinate of the center of the mobject as ``numpy.float64`` """
         return self.get_coord(1, direction)
 
     def get_z(self, direction=ORIGIN):
+        """Returns z coordinate of the center of the mobject as ``numpy.float64`` """
         return self.get_coord(2, direction)
 
     def get_start(self):
+        """Returns the point, where the stroke that surrounds the mobject starts.  Type is a ``numpy.ndarray`` with 3 elements, e.g. [-1,0,0] """
```
Comment (kolibril13) [inline manim/mobject/mobject.py:1769]: Should the part " Type is a ``numpy.ndarray`` with 3 elements, e.g. [-1,0,0] " be removed?
I think this is a bit long

**Comment by kolibril13** [comment]
Ready for review!

File: manim/mobject/mobject.py [hunk 5]
```diff
@@ -1011,9 +1011,11 @@ def scale(self, scale_factor: float, **kwargs) -> "Mobject":
             return self
 
     def rotate_about_origin(self, angle, axis=OUT, axes=[]):
+        """Rotates an mobject about the ORIGIN, which is at [0,0,0]"""
```
Comment (jsonvillanueva) [inline manim/mobject/mobject.py:1014]:
```suggestion
        """Rotates the :class:`~.Mobject` about the ORIGIN, which is at [0,0,0]."""
```

File: manim/mobject/mobject.py [hunk 6]
```diff
@@ -1011,9 +1011,11 @@ def scale(self, scale_factor: float, **kwargs) -> "Mobject":
...
         return self.rotate(angle, axis, about_point=ORIGIN)
 
     def rotate(self, angle, axis=OUT, **kwargs):
+        """Rotates an mobject about a certain point"""
```
Comment (jsonvillanueva) [inline manim/mobject/mobject.py:1018]:
```suggestion
        """Rotates the :class:`~.Mobject` about a certain point."""
```

File: manim/mobject/mobject.py [hunk 7]
```diff
@@ -1415,6 +1436,7 @@ def space_out_submobjects(self, factor=1.5, **kwargs):
     def move_to(
         self, point_or_mobject, aligned_edge=ORIGIN, coor_mask=np.array([1, 1, 1])
     ):
+        """Move center of an mobject to certain coordinate"""
```
Comment (jsonvillanueva) [inline manim/mobject/mobject.py:1439]:
```suggestion
        """Move center of the :class:`~.Mobject` to certain coordinate."""
```

File: manim/mobject/mobject.py [hunk 8]
```diff
@@ -1590,6 +1612,7 @@ def get_color(self):
     ##
 
     def save_state(self):
+        """save th current state (position, color & size). Can be restored with ``restore``"""
```
Comment (jsonvillanueva) [inline manim/mobject/mobject.py:1615]:
```suggestion
        """Save the current state (position, color & size). Can be restored with :meth:`~.Mobject.restore`."""
```

File: manim/mobject/mobject.py [hunk 9]
```diff
@@ -1598,6 +1621,7 @@ def save_state(self):
         return self
 
     def restore(self):
+        """restores the state that was previously saved with ``saved_state``  """
```
Comment (jsonvillanueva) [inline manim/mobject/mobject.py:1624]:
```suggestion
        """Restores the state that was previously saved with :meth:`~.Mobject.saved_state`."""
```

File: manim/mobject/mobject.py [hunk 10]
```diff
@@ -1687,6 +1711,7 @@ def get_corner(self, direction):
         return self.get_critical_point(direction)
 
     def get_center(self):
+        """Center coordinates the mobject"""
```
Comment (jsonvillanueva) [inline manim/mobject/mobject.py]:
```suggestion
        """Center coordinate of the :class:`~.Mobject`."""
```

File: manim/mobject/mobject.py [hunk 11]
```diff
@@ -1698,15 +1723,19 @@ def get_boundary_point(self, direction):
         return all_points[index]
 
     def get_top(self):
+        """Top coordinates the mobject"""
```
Comment (jsonvillanueva) [inline manim/mobject/mobject.py]:
```suggestion
        """Top center coordinate of the :class:`~.Mobject`."""
```

File: manim/mobject/mobject.py [hunk 12]
```diff
@@ -1698,15 +1723,19 @@ def get_boundary_point(self, direction):
...
         return self.get_edge_center(UP)
 
     def get_bottom(self):
+        """Bottom coordinates the mobject"""
```
Comment (jsonvillanueva) [inline manim/mobject/mobject.py]:
```suggestion
        """Bottom center coordinate of the :class:`~.Mobject`."""
```

File: manim/mobject/mobject.py [hunk 13]
```diff
@@ -1698,15 +1723,19 @@ def get_boundary_point(self, direction):
...
         return self.get_edge_center(DOWN)
 
     def get_right(self):
+        """RIGHT coordinates the mobject"""
```
Comment (jsonvillanueva) [inline manim/mobject/mobject.py]:
```suggestion
        """Right center coordinate of the :class:`~.Mobject`."""
```

File: manim/mobject/mobject.py [hunk 14]
```diff
@@ -1698,15 +1723,19 @@ def get_boundary_point(self, direction):
...
         return self.get_edge_center(RIGHT)
 
     def get_left(self):
+        """LEFT coordinates the mobject"""
```
Comment (jsonvillanueva) [inline manim/mobject/mobject.py]:
```suggestion
        """Left center coordinate of the :class:`~.Mobject`."""
```

File: manim/mobject/mobject.py (hunk 4 above)
Comment (jsonvillanueva) [reply manim/mobject/mobject.py:1769]: Yes. It should be under a Return section and should also use type hints.

Also this should be done for the other methods you've documented as well (but particularly get_end, and get_start_and_end)

File: manim/mobject/mobject.py [hunk 15]
```diff
@@ -1770,6 +1805,7 @@ def get_z_index_reference_point(self):
         return z_index_group.get_center()
 
     def has_points(self):
+        """Check if mobject contains points"""
```
Comment (jsonvillanueva) [inline manim/mobject/mobject.py:1808]:
```suggestion
         """Check if :class:`~.Mobject` contains points."""
```

File: manim/mobject/mobject.py [hunk 16]
```diff
@@ -1398,12 +1416,15 @@ def set_coord(self, value, dim, direction=ORIGIN):
...
     def set_x(self, x, direction=ORIGIN):
+        """Set x value of the center of the mobject (``int`` or ``float``)"""
         return self.set_coord(x, 0, direction)
 
     def set_y(self, y, direction=ORIGIN):
+        """Set y value of the center of the mobject (``int`` or ``float``)"""
         return self.set_coord(y, 1, direction)
 
     def set_z(self, z, direction=ORIGIN):
+        """Set z value of the center of the mobject (``int`` or ``float``)"""
```
Comment (jsonvillanueva) [inline manim/mobject/mobject.py:1427]: :class:`~.Mobject`, ending period

**Review by jsonvillanueva — CHANGES_REQUESTED** [review]
Mostly small things. Periods at the ends of sentences, some references to :class:\`~.Mobject\`. Also, some references to :meth:\`~.Mobject.xxxxxxxx\` should be used instead of \``xxxxxxxx\``. Technically this documentation should have type hints as well as we've been enforcing that.

**Comment by kolibril13** [comment]
Thanks for the review!
Should there really be a reference to all mobjects?
I think is not really necessary, as this is the page of mobject.
Also, it is also not there in already existing docstrings of other methods in mobject:
image
However, if this is the better approach, I am fine with doing it that way.

File: manim/mobject/mobject.py [hunk 17]
```diff
@@ -1398,12 +1416,15 @@ def set_coord(self, value, dim, direction=ORIGIN):
         return self
 
     def set_x(self, x, direction=ORIGIN):
+        """Set x value of the center of the mobject (``int`` or ``float``)"""
```
Comment (kolibril13) [inline manim/mobject/mobject.py:1419]:
```suggestion
        """Set x value of the center of the :class:`~.Mobject` (``int`` or ``float``)"""
```

File: manim/mobject/mobject.py [hunk 18]
```diff
@@ -1398,12 +1416,15 @@ def set_coord(self, value, dim, direction=ORIGIN):
...
         return self.set_coord(x, 0, direction)
 
     def set_y(self, y, direction=ORIGIN):
+        """Set y value of the center of the mobject (``int`` or ``float``)"""
```
Comment (kolibril13) [inline manim/mobject/mobject.py:1423]:
```suggestion
        """Set y value of the center of the :class:`~.Mobject` (``int`` or ``float``)"""
```

File: manim/mobject/mobject.py (hunk 16 above)
Comment (kolibril13) [inline manim/mobject/mobject.py:1427]:
```suggestion
        """Set z value of the center of the :class:`~.Mobject` (``int`` or ``float``)"""
```

File: manim/mobject/mobject.py (hunk 4 above)
Comment (kolibril13) [reply manim/mobject/mobject.py:1769]: okay!
```suggestion
        """Returns the point, where the stroke that surrounds the :class:`~.Mobject` starts. 
```

File: manim/mobject/mobject.py [hunk 19]
```diff
@@ -1721,33 +1750,39 @@ def length_over_dim(self, dim):
...
     def get_start(self):
+        """Returns the point, where the stroke that surrounds the :class:`~.Mobject` starts. 
         self.throw_error_if_no_points()
         if config["use_opengl_renderer"]:
             return np.array(self.data["points"][0])
         else:
             return np.array(self.points[0])
 
     def get_end(self):
+        """Returns the point, where the stroke that surrounds the mobject ends.  Type is a ``numpy.ndarray`` with 3 elements, e.g. [+1,0,0] """
```
Comment (kolibril13) [inline manim/mobject/mobject.py:1777]:
```suggestion
        """Returns the point, where the stroke that surrounds the :class:`~.Mobject` ends.  
```

File: manim/mobject/mobject.py [hunk 20]
```diff
@@ -1687,6 +1711,7 @@ def get_corner(self, direction):
         return self.get_critical_point(direction)
 
     def get_center(self):
+        """Get center coordinates"""
```
Comment (kolibril13) [inline manim/mobject/mobject.py:1714]: @jsonvillanueva : you were suggesting "Center coordinate of the :class:`~.Mobject`." But I think Get center coordinates will do

File: manim/mobject/mobject.py [hunk 21]
```diff
@@ -1721,33 +1750,39 @@ def length_over_dim(self, dim):
...
     def get_start(self):
+        """Returns the point, where the stroke that surrounds the :class:`~.Mobject` starts. 
         self.throw_error_if_no_points()
         if config["use_opengl_renderer"]:
             return np.array(self.data["points"][0])
         else:
             return np.array(self.points[0])
 
     def get_end(self):
+        """Returns the point, where the stroke that surrounds the :class:`~.Mobject` ends.
```
Comment (kolibril13) [inline manim/mobject/mobject.py:1777]:
```suggestion
        """Returns the point, where the stroke that surrounds the :class:`~.Mobject` ends."""
```

File: manim/mobject/mobject.py [hunk 22]
```diff
@@ -1721,33 +1750,39 @@ def length_over_dim(self, dim):
...
     def get_y(self, direction=ORIGIN):
+        """Returns y coordinate of the center of the mobject as ``numpy.float64`` """
         return self.get_coord(1, direction)
 
     def get_z(self, direction=ORIGIN):
+        """Returns z coordinate of the center of the mobject as ``numpy.float64`` """
         return self.get_coord(2, direction)
 
     def get_start(self):
+        """Returns the point, where the stroke that surrounds the :class:`~.Mobject` starts.
```
Comment (kolibril13) [inline manim/mobject/mobject.py:1769]:
```suggestion
        """Returns the point, where the stroke that surrounds the :class:`~.Mobject` starts."""
```

**Review by huguesdevimeux — APPROVED** [review]

File: manim/mobject/mobject.py [hunk 23]
```diff
@@ -1770,45 +1806,46 @@ def get_z_index_reference_point(self):
...
-    def match_y(self, mobject, direction=ORIGIN):
+    def match_y(self, mobject:"Mobject", direction=ORIGIN):
         return self.match_coord(mobject, 1, direction)
 
-    def match_z(self, mobject, direction=ORIGIN):
+    def match_z(self, mobject:"Mobject", direction=ORIGIN):
         return self.match_coord(mobject, 2, direction)
 
-    def align_to(self, mobject_or_point, direction=ORIGIN, alignment_vect=UP):
+    def align_to(self, mobject_or_point:Union["Mobject",np.ndarray, List], direction=ORIGIN, alignment_vect=UP):
```
Comment (kolibril13) [inline manim/mobject/mobject.py:1848]: As I am using type hints the first time, I am not sure if this is the best way to add them in this way `Union["Mobject",np.ndarray, List]`, or if it is even possible to specify that the list/array needs to have exactly 3 entries.

**Comment by jsonvillanueva** [comment]
> 
> 
> Thanks for the review!
> Should there really be a reference to all mobjects?
> I think is not really necessary, as this is the page of mobject.
> Also, it is also not there in already existing docstrings of other methods in mobject:
> image
> However, if this is the better approach, I am fine with doing it that way.

This is more about completeness, but essentially these links will take the reader back to the top of the page that they're already on. Similarly, if you mention a method/function, they should also be linked to properly (even if they're on the same page) because the user can quickly jump to that section if they want to click on it and go to it.

File: manim/mobject/mobject.py (hunk 20 above)
Comment (jsonvillanueva) [reply manim/mobject/mobject.py:1716]: Yes, this is fine. There's no *real* need to mention Mobject here, but if you do, it's better to have it linked/clickable

**Review by jsonvillanueva — APPROVED** [review]

**Pull request merged**